### DISCUSSION TITLE
Does renovate CHANGELOG detection support multiple nuget packages hosted in one repository?

### DISCUSSION THREAD

**Discussion by krajek** [renovatebot/renovate#10324] [Request Help]
My company has a single repository that contains many of our internal NuGet packages.
Additionally, some microservices also publish one or more packages (for nuget/npm/etc some or all) from one repository.

Unfortunately, it seems that CHANGELOG is looked up only at the root of the default branch, which breaks CHANGELOG detection. 
**Am I right? Is this by design or maybe it can be improved?**

Presumably relevant source code:
https://github.com/renovatebot/renovate/blob/e0488a3ebbd8ddc5bbe3d1a303ec4cf31ad6cf7f/lib/workers/pr/changelog/github/index.ts#L43-L87

What we would expect is either to take changelog from the package itself(it is contained within the nuget package) or from URL read from `<RepositoryUrl>` or `<ProjectUrl>`.

**Comment by rarkins** [comment]
Renovate doesn't currently read changelogs from registries, but it's something which could be added as a feature request if you can document the relevant approach for Nuget to begin with.

There's a little bit of existing logic for trying to find JS packages within a monorepo but not for other languages. What does your repository directory/file structure look like (only caring about packages and release notes of course). e.g. do you have `./packages/foo/CHANGEOG.md` and `./packages/bar/CHANGELOG.md`?

**Reply by krajek** [reply]
Thanks for the reply @rarkins.

The simplest solution(or at least the first step) would be to display the `<PackageReleaseNotes>` value somewhere in the PR. It is a standard field for `.csproj` although I have to admit that even Visual Studio does not display it in the built-in NuGet package manager. Nonetheless, that would be a way to easily support "monorepo" scenarios.

That said, as far as I understand, such change would require some refactorings since `getReleaseNotesMd` is a datasource-agnostic function. It does not accept any datasource related information as parameters. I can code typescript so I could even provide a PR, I would just need a confirmation that the idea is a sound one.

On a side note, nobody seems to support these monorepo scenarios. Neither GitHub Releases, nor Dependabot, nor other tools. Everybody seems to assume that there is one package per repo, which often is impractical.

**Reply by rarkins** [reply]
I would like to first see the official docs for `PackageReleaseNotes` as well as an example monorepo using it. Without those I couldn't really say for sure if it's a sound idea and wouldn't want you to work on something we ultimately can't accept.

e.g. I'd like to understand if they're embedded per-release or if it's a large `PackageReleaseNotes` including every release? And is that field made available through the standard nuget API?

**Reply by krajek** [reply]
nuspec doc for releaseNotes: https://docs.microsoft.com/en-us/nuget/reference/nuspec#releasenotes
nuget pack documentation, mentions PackageReleaseNotes: https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets

Example repositories:
https://github.com/fluentassertions/fluentassertions/blob/b2f8004885ad41fbdd16702e24916c473a7a2d5a/Src/FluentAssertions/FluentAssertions.csproj#L28-L30
https://github.com/RehanSaeed/Serilog.Exceptions/blob/98777933534c59069a53faf4a77c0a990cfc3ec8/Directory.Build.props#L19-L21
https://github.com/jsakamoto/EntityFrameworkCore.DecimalAttribute/blob/d518834d4180f83d67ae7cd4f9c5ada844d80324/EFCore.DecimalAttribute/EFCore.DecimalAttribute.csproj#L13-L15
release.props usage: 
https://github.com/andrewlock/PwnedPasswords/blob/2e3f117bcd3939cd53e25ab4e1ca15cea918398e/releasenotes.props#L3-L15

nuget v2 example, contains `<ReleaseNotes>` https://www.nuget.org/api/v2/FindPackagesById()?id=%27Serilog.Exceptions%27

**Reply by krajek** [reply]
After more research, I can say that even though `<PackageReleaseNotes>` is standard and well supported by `dotnet` and `nuget` themselves, they are not ubiquitously filled by package creators. Those who do, sometimes paste a URL with a path and sometimes paste full CHANGELOG, either manually or automatically.

Still, it seems like if someone is filling the property, we can treat its value as a meaningful source. If the value is empty then we could proceed to other sources like looking up CHANGELOGs in the root of the repo or in Github Releases.

**Reply by rarkins** [reply]
Are any of these available via the nuget API? If for example the `nuspec` file with `releaseNotes` is the downloadable tarball then that's not something we currently access or extract from. Running `curl` on `https://www.nuget.org/api/v2/FindPackagesById()?id=%27Serilog.Exceptions%27` doesn't work for me.
